vbfml/scripts/err_analysis

```diff
@@ -783,5 +826,66 @@ def plot_correl(ctx, input_dir: str) -> None:
         f.write(table_print)
 
 
+@cli.command()
+@click.pass_context
+@click.option(
+    "-i",
+    "--input-dir",
+    required=True,
+    help="path to the prediction directory in a model",
+)
+@click.option(
+    "-c", "--cut-variable", required=True, help="The quantity to cut on the threshold."
+)
+@click.option(
+    "-t", "--threshold", type=float, required=True, help="The threshold/cut value."
+)
+@click.option(
+    "-p",
+    "--plot-variable",
+    default="score",
+    required=True,
+    help="The variable to plot the distribution after the cut",
+)
+def dist_cut(
+    ctx, input_dir: str, cut_variable: str, threshold: float, plot_variable: str
+) -> None:
+    """
+    Apply a cut on the  cut-variable and plot the distribution of the plot-variable after the cut
+    """
+
+    # load data
+    features, predictions, _, labels = load_cache(input_dir)
+
+    # adapt variable if "score" is wanted, because "score" is not in feature
+    read_plot_variable= "mjj" if (plot_variable == "score") else plot_variable
+    read_cut_variable= "mjj" if (cut_variable == "score") else cut_variable
+
+    d = pd.DataFrame(
+        {
+            read_cut_variable: features[read_cut_variable],
+            read_plot_variable: features[read_plot_variable],
+            "labels": labels,
+            "weights": features["weight"],
+            "score": predictions,
+        }
+    )
+
+    # apply cut
+    d = d[d[cut_variable] > threshold]
+
+    # save the plot
+    outdir = pjoin(input_dir, "distribution_cut")
+    if not os.path.exists(outdir):
+        os.makedirs(outdir)
+
+    plot_histograms_for_each_label(
+        data=d,
+        variable=plot_variable,
+        outdir=outdir,
+        save_name=f"_cut_{cut_variable}_{threshold}",
+    )
+
+
```

Comment:
So to build on our earlier discussion, I'm thinking how practical would it be to use this function to plot "EWK vs VBF" or "QCD+EWK vs VBF", depending on a command-line argument.

I see that, it actually takes data from the cache created by the earlier `predict` function, where the dataset configuration is set (i.e. what is background and what is signal). So we'll need to run `predict` first, and then this function again to change what we're plotting, does that sound correct?

Reply:
I'm thinking we can actually make `plot_histograms_for_each_label` function more flexible, so that we can plot a 3-class histogram if we want to, but I need to think about that a little bit. Currently I see that it is hard-coded to plot only `0` label vs `1` label, maybe we can provide it with a `labels` argument somehow, and use a `for` loop to plot the histograms on top of each other. This way, we can provide 2 or 3 sets of data and plot them with the same function, but I'll need to think about what exactly to provide to this function. Does that sound reasonable to you? Let me know if you have any suggestions

Reply:
Yes exactly that's the problem that everything need to come from predict-cache and the predict function then, which is not really handy for what we want to do.
1. We can use the predict on all validation all kind of events and so we have QCD + EWK backgrounds at the same place, and after plot the 3 separately anyhow (but still need to use predict, even if we don't look at the predictions for these histograms)
2. we can think of a new function totally independent that does all the job: read the good backgrounds + produce them labels and plot the histo according to a given cut

Reply:
Thanks, I'm thinking this can be achieved by introducing a flexibility to `MultiDatasetSequence` object. Currently the buffer of this object is holding a `pd.DataFrame`, one thing we can try is to add a column to this DataFrame, so that we store whether the given event is `VBF H`, `EWK V` or `QCD V`. This information can then be read from the `features` dictionary, and can be used for the plotting.

Reply:
I will try to include this implementation in the `MultiDatasetSequence` object, and I can push a commit to this branch so it's included with this PR. And once that's in place, we can update this function to accommodate that change, if that sounds good.

Reply:
Hey @pmayenco, I just made this update in [0c5e9d9](https://github.com/alpakpinar/vbfml/pull/39/commits/0c5e9d92a506f85f6c99306a3078ccfa5940b0ce) (with a few more commits to fix the unit tests). Now the `features` dictionary should contain a `dataset_label`, and this will be one of the following: `qcd_v`, `ewk_v`, `vbf_h`. You can see how they're assigned [here](https://github.com/alpakpinar/vbfml/blob/pierre_branch_17_05_22/vbfml/input/sequences.py#L34-L50).

Can you try to use this new feature here in the plotting script? So instead of plotting data with `label==0` and `label==1`, you can try to plot data according to the `dataset_label`. Maybe you can pass which `dataset_label`s to plot to the `plot_histograms_for_each_label` function here, and in the function, you can do a `for` loop over them to plot them on the same figure. Let me know what you think, thanks! 

And since I've committed on this branch, please update your branch before committing: `git pull origin pierre_branch_17_05_22`

Reply:
okay this is done 
I checked that I gave me the exact same histogram as the one I made with my other dummy function for the presentation
- So you have to predict first and then cut and plot, if you predict only on ewk and vbfh, it will plot only the 2 histo
- you don't have to set a cut variable or threshold, so if you don't it just plot the distribution without cut

Reply:
Great, thanks! Do you think if we can make the choice of datasets to plot more flexible? I was thinking instead of hard-coding the list of datasets we want to in the `plot_histograms_for_each_label` function, we can make it a command line argument to `cut_and_plot` somehow (maybe the default being all three of them), and then feeding this list to the `plot_histograms_for_each_label` function

Reply:
For the `cut_and_plot` function, we can maybe add a `--datasets` command line option, and following the documentation [here](https://click.palletsprojects.com/en/7.x/options/#multiple-options), something like this might work:

```python
@click.option("--datasets", multiple=True, default=["qcd_v","ewk_v","vbf_h"])

# And then pass the datasets list to the plot_histograms function
```

Reply:
sorry I missed your last comment but that is basically what I did

but the only problem is that it is gonna save on the same name with or without specifying the dataset, I don't know how to solve this yet

Reply:
Thanks, looks good. Hmm yeah so the PDF file doesn't automatically change, I think for now we can leave it to the user  to specify a proper name for each PDF file, if you think that sounds good.

Reply:
done